File: py/py_tests/data_structure_tests/BinarySearchTreeTest.py
import unittest
from py.data_structures.BinarySearchTree import BinarySearchTree

class Tests(unittest.TestCase):
    def testInsertion(self):
        bst = BinarySearchTree(3)
        for val in range(8):
            bst.insertion(val)
        
        self.assertEqual(validateBST(bst), True) 
    
    def testLookup(self):
        bst = BinarySearchTree(4)
        ints = [1, 2, 3, 5, 6, 7, 8]
        for val in ints:
            bst.insertion(val)
        
        for val in ints:
            self.assertEqual(bst.lookup(val).val, val) 


def validateBST(bst):
    return validateBSTHelper(bst.root, float('-inf'), float('inf'))

def validateBSTHelper(node, lowBound, highBound):
    if not node:
        return True 
    if not lowBound <= node.val < highBound:
        return False 
    valid_left_subtree = validateBSTHelper(node.left, lowBound, node.val)
    valid_right_subtree = validateBSTHelper(node.right, node.val, highBound)
    return valid_left_subtree and valid_right_subtree



if __name__ == "__main__":
    unittest.main() 

File: ts/src/tests/data_structure_tests/binary_search_tree.test.ts
import BinarySearchTree from '../../data structures/binary_search_tree';
import { BinaryTreeNode } from '../../data structures/utils/binary_tree';

function validateBSTHelper(
    node: BinaryTreeNode | null,
    lowBound: number,
    highBound: number,
): boolean {
    if (!node) {
        return true;
    }
    if (!(node.val >= lowBound && node.val < highBound)) {
        return false;
    }
    const validSubtreeLeft = validateBSTHelper(node.left, lowBound, node.val);
    const validSubtreeRight = validateBSTHelper(
        node.right,
        node.val,
        highBound,
    );
    return validSubtreeLeft && validSubtreeRight;
}

function validateBST(bst: BinarySearchTree): boolean {
    return validateBSTHelper(
        bst.root,
        Number.NEGATIVE_INFINITY,
        Number.POSITIVE_INFINITY,
    );
}
describe('Tests testing binary search tree', () => {
    test('testing bst insertion method', () => {
        const bst = new BinarySearchTree(4);
        const vals = [1, 2, 3, 4, 5, 6, 7, 8];
        vals.forEach((val) => {
            bst.insert(val);
        });
        expect(validateBST(bst)).toEqual(true);
    });

    test('testing bst lookup method', () => {
        const bst = new BinarySearchTree(4);
        const vals = [1, 2, 3, 4, 5, 6, 7, 8];
        vals.forEach((val) => {
            bst.insert(val);
        });

        vals.forEach((val) => {
            const returnNode = bst.lookup(val);
            if (returnNode != null) {
                expect(returnNode.val).toEqual(val);
            }
        });
    });

    test('testing bst delete method', () => {
        const bst = new BinarySearchTree(4);
        const vals = [1, 2, 3, 4, 5, 6, 7, 8];

        vals.forEach((val) => {
            bst.insert(val);
        });

        vals.forEach((val) => {
            const returnNode = bst.lookup(val);
            if (returnNode != null) {
                expect(returnNode.val).toEqual(val);
            }
            bst.delete(val);
            if (val !== 4) {
                expect(bst.lookup(val)).toEqual(null);
            }
            const lookedUpNode = bst.lookup(val);
            if (lookedUpNode != null) {
                expect(lookedUpNode.val).toEqual(4);
            }
        });
    });
});
